sdk/core/switcher/entry.S

```diff
@@ -181,8 +180,24 @@ compartment_switcher_entry:
 	cgetbase           s1, csp
 	csetaddr           csp, csp, s1
 	sub                s1, s0, s1
-	csetboundsexact    csp, csp, s1
-	zero_stack         sp, s0, gp
+	csetboundsexact    ct2, csp, s1
+	csetaddr           csp, ct2, s0
+#ifdef CONFIG_MSLWM
+	// read and align the stack low water mark
+	csrr               gp, 0xbc1 // mslwm
+	and                gp, gp, ~0xf
+	// skip zeroing if low water mark >= stack poitner
+	bge                t2, sp, after_zero
```

Comment:
We probably need to zero the whole thing if the address is not in bounds.  I think that is not needed on the return path though.

Reply:
I think it might be better to fault and unwind to the caller. Here are some scenarios (`hwm` == high water mark):

1. `csp.addr > csp.top` this is possible before this PR. It could happen only if the caller messes with `csp`. It would trap in `csetboundsexact`. According to the comment this will cause a forced unwind because it is after the update to the tstack offset. 
3. `hwm > csp.addr` this can happen in normal operation and is explicitly handled by not zeroing anything (we maintain an invariant that the stack between base and `hwm` is always zero).
4. `hwm > csp.top` this could happen if the caller reduces the length of csp before calling, which would be weird. What happens depends on the value of `csp.addr`. If `hwm > csp.addr` then we skip zeroing as above. This is harmless and the call proceeds with the smaller stack. If `hwm <= csp.addr` we will proceed to zeroing but will trap in `zero_stack`, either because `ct2` became unrepresentable on `csetaddr` and had its tag cleared or because we eventually try to zero above `csp.top`. This trap should cause a forced unwind.
5. `csp.addr < csp.base` the current capability encoding makes this impossible. If `csp.addr` is moved below `csp.base` it's tag will be cleared and this will be detected in `check_compartment_stack_integrity`
6. high water mark `< csp.base` This is possible only if the caller moves `csp.base`, which would be weird. In this case `csetaddr` would make `ct2` unrepresentable, which would clear the tag causing a trap in `zero_stack` again.

If traps are handled as the described in the comment I don't think any special handling is required.

Reply:
I need to think about this a bit more.  It is fine for the caller to restrict the stack before calling (which it might choose to do to avoid some stack clearing on the way out if it knows the callee doesn't need much stack).  I think we check the permissions on csp, so there's no way for it to be anything other than something derived from a valid csp here.

Reply:
Thinking a bit more, case 3 is interesting because the correct behaviour here is to zero all of csp but *not* update the SHWM register.  Possibly we should disallow this entirely.

Reply:
Updated so that we now just force unwind if we fault anywhere in the switcher.  This is simpler than special casing a load of things.